src/golang/lib/engine/authenticate.go

```diff
@@ -43,11 +49,69 @@ func AuthenticateLambdaConfig(ctx context.Context, authConf auth.Config) error {
 		lambda_utils.SnowflakeConnectorType,
 	}
 
-	for _, functionType := range functionsToShip {
-		err := lambda_utils.CreateLambdaFunction(functionType, lambdaConf.RoleArn)
-		if err != nil {
-			return errors.Wrap(err, "Unable to Create Lambda Function")
-		}
+	// Run authentication only once since the credentials will be recorded.
+	err = lambda_utils.AuthenticateDockerToECR()
+	if err != nil {
+		return errors.Wrap(err, "Unable to authenticate Lambda Function.")
+	}
+
+	errGroup, errGroupCtx := errgroup.WithContext(ctx)
+	// Pull images on a concurrency of "MaxConcurrentDownload".
+	pullImageChannel := make(chan lambda_utils.LambdaFunctionType, len(functionsToShip))
+	defer close(pullImageChannel)
+	pushImageChannel := make(chan lambda_utils.LambdaFunctionType, len(functionsToShip))
+	defer close(pushImageChannel)
+	lambda_utils.AddFunctionTypeToChannel(functionsToShip[:], pullImageChannel)
+	for i := 0; i < MaxConcurrentDownload; i++ {
+		errGroup.Go(func() error {
+			for {
+				select {
+				case functionType := <-pullImageChannel:
+					lambdaFunctionType := functionType
+					err := lambda_utils.PullImageFromECR(lambdaFunctionType)
+					if err != nil {
+						return err
+					}
+					pushImageChannel <- functionType
+				case <-errGroupCtx.Done():
+					return errGroupCtx.Err()
+				default:
+					return nil
```

Comment:
Can you add a comment here that this case should only be hit when the `pullImageChannel` is empty?